modules/stochastic_tools/include/surrogates/SurrogateTrainer.h

```diff
@@ -70,11 +85,201 @@ SurrogateTrainer::declareModelData(const std::string & data_name, const T & valu
+
+  // TRAINING_DATA_END
+
+  /// Sampler being used for training
+  Sampler & _sampler;
+
+  /// During training loop, this is the row index of the data
+  dof_id_type _row;
+  /// During training loop, this is the local row index of the data
+  dof_id_type _local_row;
+
+private:
+  /*
+   * Called at the beginning of execute() to make sure values are set properly
+   */
+  void checkIntegrity() const;
+
+  /// Sampler data for the current row
+  std::vector<Real> _row_data;
+
+  /// Vector of reporter names and their corresponding values (to be filled by getTrainingData)
+  std::unordered_map<ReporterName, std::shared_ptr<TrainingDataBase>> _training_data;
+};
+
+template <typename T>
+const T &
+SurrogateTrainer::getTrainingData(const std::string & param)
+{
+  return getTrainingData<T>(getParam<ReporterName>(param));
+}
+
+template <typename T>
+std::vector<const T *>
+SurrogateTrainer::getTrainingDataVector(const std::string & param, bool use_sampler_default)
+{
+  std::vector<const T *> vals;
+  if (isParamValid(param) || !use_sampler_default)
+  {
+    for (const ReporterName & rname : getParam<std::vector<ReporterName>>(param))
+      vals.push_back(&getTrainingData<T>(rname));
+  }
+  else
+  {
+    if (!std::is_same<T, Real>::value)
+      mooseError("Training data values must be type Real if taken from sampler.");
+    for (const Real & col : _row_data)
+      vals.push_back(&col);
+  }
+
+  return vals;
+}
+
+template <typename T>
+const T &
+SurrogateTrainer::getTrainingData(const ReporterName & rname)
+{
+  if (rname.getObjectName() == "sampler")
+  {
+    if (!std::is_same<T, Real>::value)
+      mooseError("Training data values must be type Real if taken from sampler.");
+
+    std::size_t ind = rname.getValueName().find("_");
+    if (ind != std::string::npos)
+    {
+      unsigned int col = MooseUtils::stringToInteger(rname.getValueName().substr(ind + 1), true);
+      if (col >= _sampler.getNumberOfCols())
+        mooseError("Training data from sampler ",
+                   rname,
+                   " is requesting an index greater than the number of sampler columns (",
+                   _sampler.getNumberOfCols(),
+                   ").");
+      return _row_data[col];
+    }
+    else
+      mooseError("Predictor data ", rname, " from sampler is in the wrong format.");
+  }
+  else
+  {
+    auto it = _training_data.find(rname);
+    if (it != _training_data.end())
+    {
+      auto data = std::dynamic_pointer_cast<TrainingData<T>>(it->second);
+      if (!data)
+        mooseError("Reporter value ", rname, " already exists but is of different type.");
+      return data->get();
+    }
+    else
+    {
+      const std::vector<T> & rval = getReporterValueByName<std::vector<T>>(rname);
+      _training_data[rname] = std::make_shared<TrainingData<T>>(rval);
+      return std::dynamic_pointer_cast<TrainingData<T>>(_training_data[rname])->get();
+    }
+  }
+}
+
+class TrainingDataBase
```

Comment:
Can/should the _is_distributed be passed in the constructor and then be const?

Reply:
Unfortunately, the function calling the constructor can be called before the reporter value has been declared. As such we don't know if the value is distributed or not, so we have to check this later in `initialize()`.